File: data_structures/Java/Queue.java
package Java;

public class Queue<T> {
    
    private Node front_pointer;
    private Node back_pointer;
    private int size;

    class Node {
        private T value;
        private Node next;

        Node(T value) {
            this.value = value;
            next = null;
        }

        public T getValue() {
            return value;
        }

        public Node getNext() {
            return next;
        }

        public void setNext(Node node) {
            next = node;
        }

    }

    Queue() {
        Node front_pointer = null;
        Node back_pointer = null;
        int size = 0;
    }

    public static void main(String[] args) {
        Integer[] data = {25, 39, 46, 20, 3, 23, 35, 32, 8, 16, 12, 33, 36, 31, 49, 41, 2, 4, 6, 9, 17, 34, 24, 13, 0, 42, 37, 40, 44, 30};
        Queue<Integer> queue = new Queue<Integer>();
        for (int d : data) {
            queue.enqueue(d);
        }
        System.out.println(queue.get_head());
        System.out.println(queue.get_tail());
        do {
            System.out.println(queue.dequeue());
        } 
        while (queue.size>0);
    }

    public T get_head() {
        return front_pointer.value;
    }

    public T get_tail() {
        return back_pointer.value;
    }

    public void enqueue(T data) {
        Node new_node = new Node(data);
        //Check if queue is empty
        if (back_pointer == null) {
            front_pointer = new_node;
        } else {
            back_pointer.setNext(new_node);
        }

        back_pointer = new_node;
        size ++;
    }

    public T dequeue() {
        //Check for queue underflow
        if (front_pointer != null) {
            T popped = front_pointer.value;
            front_pointer = front_pointer.next;
            size --;

            //Check if last item was popped
            if (front_pointer == null) {
                back_pointer = null;
            }
            return popped;
        } else {
            return null;
        }
    }
}

File: data_structures/Python/Queue.py
class Queue:
    
    class Node:
        def __init__(self,data):
            self.data = data
            self.pointer = None
    
    def __init__(self):
        self.front_pointer = None
        self.back_pointer = None
        self.size = 0

    def enqueue(self,data):
        ##Check for queue overflow
        try:
            new_node = Queue.Node(data)
            ##Check if queue is empty
            if self.back_pointer == None:
                self.front_pointer = new_node
            else:
                self.back_pointer.pointer = new_node

            self.back_pointer = new_node
            self.size += 1
            return True

        except:
            return False

    def dequeue(self):
        ##Check for queue underflow
        if self.front_pointer != None:
            popped = self.front_pointer.data
            self.front_pointer = self.front_pointer.pointer
            
            ##Check if last item was popped
            if self.front_pointer == None:
                self.back_pointer = None
            self.size -= 1
            return popped

        else:
            return None
        
    def peek(self):
        ##Check for queue underflow
        if self.front_pointer != None:
            return self.front_pointer.data
        else:
            return None
        
    def empty(self):
        self.front_pointer = self.back_pointer = None



def show():
    items = ["Florida","Georgia","Delaware","Alabama","California"]
    q = Queue()
    for i in items:
        q.enqueue(i)
        print(q.size)

    item = ''
    while item != None:
        print(item)
        item = q.dequeue()
        
    
##show()
